python/ray/tests/test_failure.py

```diff
@@ -727,24 +728,44 @@ def test_warning_for_too_many_nested_tasks(shutdown_only):
     ray.init(num_cpus=num_cpus)
     p = init_error_pubsub()
 
+    remote_wait = Semaphore.remote(value=0)
+    nested_wait = Semaphore.remote(value=0)
+
+    ray.get([
+        remote_wait.locked.remote(),
+        nested_wait.locked.remote(),
+    ])
+
     @ray.remote
     def f():
         time.sleep(1000)
         return 1
 
     @ray.remote
-    def h():
-        time.sleep(1)
+    def h(nested_waits):
+        nested_wait.release.remote()
+        ray.get(nested_waits)
         ray.get(f.remote())
 
     @ray.remote
-    def g():
+    def g(remote_waits, nested_waits):
         # Sleep so that the f tasks all get submitted to the scheduler after
         # the g tasks.
-        time.sleep(1)
-        ray.get(h.remote())
+        remote_wait.release.remote()
+        # wait until every lock is released.
+        ray.get(remote_waits)
+        ray.get(h.remote(nested_waits))
+
+    num_root_tasks = num_cpus * 6
```

Comment:
we can probably bump this back to 4. I made it 6 in the failed attempt to improve the test last week.